// cally/dist/cally.js
class te {
  /**
   * @type {T}
   */
  #t;
  #e = /* @__PURE__ */ new Set();
  /**
   * @param {T} current
   */
  constructor(t) {
    this.#t = t;
  }
  /**
   * @return {T}
   */
  get current() {
    return this.#t;
  }
  /**
   * @param {T} value
   */
  set current(t) {
    this.#t != t && (this.#t = t, this.#e.forEach((n) => n(t)));
  }
  /**
   * @type {import("hooks").Ref["on"]}
   */
  on(t) {
    return this.#e.add(t), () => this.#e.delete(t);
  }
}
const Pt = (e) => new te(e), Q = Symbol.for("atomico.hooks");
globalThis[Q] = globalThis[Q] || {};
let N = globalThis[Q];
const ee = Symbol.for("Atomico.suspense"), Nt = Symbol.for("Atomico.effect"), ne = Symbol.for("Atomico.layoutEffect"), kt = Symbol.for("Atomico.insertionEffect"), k = (e, t, n) => {
  const { i: s, hooks: o } = N.c, r = o[s] = o[s] || {};
  return r.value = e(r.value), r.effect = t, r.tag = n, N.c.i++, o[s].value;
}, se = (e) => k((t = Pt(e)) => t), U = () => k((e = Pt(N.c.host)) => e), Ot = () => N.c.update, oe = (e, t, n = 0) => {
  let s = {}, o = !1;
  const r = () => o, c = (i, u) => {
    for (const d in s) {
      const l = s[d];
      l.effect && l.tag === i && (l.value = l.effect(l.value, u));
    }
  };
  return { load: (i) => {
    N.c = { host: t, hooks: s, update: e, i: 0, id: n };
    let u;
    try {
      o = !1, u = i();
    } catch (d) {
      if (d !== ee)
        throw d;
      o = !0;
    } finally {
      N.c = null;
    }
    return u;
  }, cleanEffects: (i) => (c(kt, i), () => (c(ne, i), () => {
    c(Nt, i);
  })), isSuspense: r };
}, O = Symbol.for;
function At(e, t) {
  const n = e.length;
  if (n !== t.length)
    return !1;
  for (let s = 0; s < n; s++) {
    let o = e[s], r = t[s];
    if (o !== r)
      return !1;
  }
  return !0;
}
const T = (e) => typeof e == "function", R = (e) => typeof e == "object", { isArray: re } = Array, V = (e, t) => (t ? e instanceof HTMLStyleElement : !0) && "hydrate" in (e?.dataset || {});
function Rt(e, t) {
  let n;
  const s = (o) => {
    let { length: r } = o;
    for (let c = 0; c < r; c++) {
      const f = o[c];
      if (f && Array.isArray(f))
        s(f);
      else {
        const a = typeof f;
        if (f == null || a === "function" || a === "boolean")
          continue;
        a === "string" || a === "number" ? (n == null && (n = ""), n += f) : (n != null && (t(n), n = null), t(f));
      }
    }
  };
  s(e), n != null && t(n);
}
const Mt = (e, t, n) => (e.addEventListener(t, n), () => e.removeEventListener(t, n));
class Ft {
  /**
   *
   * @param {HTMLElement} target
   * @param {string} message
   * @param {string} value
   */
  constructor(t, n, s) {
    this.message = n, this.target = t, this.value = s;
  }
}
class It extends Ft {
}
class ce extends Ft {
}
const j = "Custom", ie = null, ae = { true: 1, "": 1, 1: 1 };
function le(e, t, n, s, o) {
  const {
    type: r,
    reflect: c,
    event: f,
    value: a,
    attr: i = ue(t)
  } = n?.name != j && R(n) && n != ie ? n : { type: n }, u = r?.name === j && r.map, d = a != null ? r == Function || !T(a) ? () => a : a : null;
  Object.defineProperty(e, t, {
    configurable: !0,
    /**
     * @this {import("dom").AtomicoThisInternal}
     * @param {any} newValue
     */
    set(l) {
      const m = this[t];
      d && r != Boolean && l == null && (l = d());
      const { error: p, value: y } = (u ? he : me)(
        r,
        l
      );
      if (p && y != null)
        throw new It(
          this,
          `The value defined for prop '${t}' must be of type '${r.name}'`,
          y
        );
      m != y && (this._props[t] = y ?? void 0, this.update(), f && $t(this, f), this.updated.then(() => {
        c && (this._ignoreAttr = i, fe(this, r, i, this[t]), this._ignoreAttr = null);
      }));
    },
    /**
     * @this {import("dom").AtomicoThisInternal}
     */
    get() {
      return this._props[t];
    }
  }), d && (o[t] = d()), s[i] = { prop: t, type: r };
}
const $t = (e, { type: t, base: n = CustomEvent, ...s }) => e.dispatchEvent(new n(t, s)), ue = (e) => e.replace(/([A-Z])/g, "-$1").toLowerCase(), fe = (e, t, n, s) => s == null || t == Boolean && !s ? e.removeAttribute(n) : e.setAttribute(
  n,
  t?.name === j && t?.serialize ? t?.serialize(s) : R(s) ? JSON.stringify(s) : t == Boolean ? "" : s
), de = (e, t) => e == Boolean ? !!ae[t] : e == Number ? Number(t) : e == String ? t : e == Array || e == Object ? JSON.parse(t) : e.name == j ? t : (
  // TODO: If when defining reflect the prop can also be of type string?
  new e(t)
), he = ({ map: e }, t) => {
  try {
    return { value: e(t), error: !1 };
  } catch {
    return { value: t, error: !0 };
  }
}, me = (e, t) => e == null || t == null ? { value: t, error: !1 } : e != String && t === "" ? { value: void 0, error: !1 } : e == Object || e == Array || e == Symbol ? {
  value: t,
  error: {}.toString.call(t) !== `[object ${e.name}]`
} : t instanceof e ? {
  value: t,
  error: e == Number && Number.isNaN(t.valueOf())
} : e == String || e == Number || e == Boolean ? {
  value: t,
  error: e == Number ? typeof t != "number" ? !0 : Number.isNaN(t) : e == String ? typeof t != "string" : typeof t != "boolean"
} : { value: t, error: !0 };
let ye = 0;
const pe = (e) => {
  const t = (e?.dataset || {})?.hydrate || "";
  return t || "c" + ye++;
}, Y = (e, t = HTMLElement) => {
  const n = {}, s = {}, o = "prototype" in t && t.prototype instanceof Element, r = o ? t : "base" in t ? t.base : HTMLElement, { props: c, styles: f } = o ? e : t;
  class a extends r {
    constructor() {
      super(), this._setup(), this._render = () => e({ ...this._props });
      for (const u in s)
        this[u] = s[u];
    }
    /**
     * @returns {import("core").Sheets[]}
     */
    static get styles() {
      return [super.styles, f];
    }
    async _setup() {
      if (this._props)
        return;
      this._props = {};
      let u, d;
      this.mounted = new Promise(
        (E) => this.mount = () => {
          E(), u != this.parentNode && (d != u ? this.unmounted.then(this.update) : this.update()), u = this.parentNode;
        }
      ), this.unmounted = new Promise(
        (E) => this.unmount = () => {
          E(), (u != this.parentNode || !this.isConnected) && (l.cleanEffects(!0)()(), d = this.parentNode, u = null);
        }
      ), this.symbolId = this.symbolId || Symbol(), this.symbolIdParent = Symbol();
      const l = oe(
        () => this.update(),
        this,
        pe(this)
      );
      let m, p = !0;
      const y = V(this);
      this.update = () => (m || (m = !0, this.updated = (this.updated || this.mounted).then(() => {
        try {
          const E = l.load(this._render), h = l.cleanEffects();
          return E && //@ts-ignore
          E.render(this, this.symbolId, y), m = !1, p && !l.isSuspense() && (p = !1, !y && be(this)), h();
        } finally {
          m = !1;
        }
      }).then(
        /**
         * @param {import("internal/hooks.js").CleanUseEffects} [cleanUseEffect]
         */
        (E) => {
          E && E();
        }
      )), this.updated), this.update();
    }
    connectedCallback() {
      this.mount(), super.connectedCallback && super.connectedCallback();
    }
    disconnectedCallback() {
      super.disconnectedCallback && super.disconnectedCallback(), this.unmount();
    }
    /**
     * @this {import("dom").AtomicoThisInternal}
     * @param {string} attr
     * @param {(string|null)} oldValue
     * @param {(string|null)} value
     */
    attributeChangedCallback(u, d, l) {
      if (n[u]) {
        if (u === this._ignoreAttr || d === l)
          return;
        const { prop: m, type: p } = n[u];
        try {
          this[m] = de(p, l);
        } catch {
          throw new ce(
            this,
            `The value defined as attr '${u}' cannot be parsed by type '${p.name}'`,
            l
          );
        }
      } else
        super.attributeChangedCallback(u, d, l);
    }
    static get props() {
      return { ...super.props, ...c };
    }
    static get observedAttributes() {
      const u = super.observedAttributes || [];
      for (const d in c)
        le(this.prototype, d, c[d], n, s);
      return Object.keys(n).concat(u);
    }
  }
  return a;
};
function be(e) {
  const { styles: t } = e.constructor, { shadowRoot: n } = e;
  if (n && t.length) {
    const s = [];
    Rt(t, (o) => {
      o && (o instanceof Element ? n.appendChild(o.cloneNode(!0)) : s.push(o));
    }), s.length && (n.adoptedStyleSheets = s);
  }
}
const Ut = (e) => (t, n) => {
  k(
    /**
     * Clean the effect hook
     * @type {import("internal/hooks.js").CollectorEffect}
     */
    ([s, o] = []) => ((o || !o) && (o && At(o, n) ? s = s || !0 : (T(s) && s(), s = null)), [s, n]),
    /**
     * @returns {any}
     */
    ([s, o], r) => r ? (T(s) && s(), []) : [s || t(), o],
    e
  );
}, ot = Ut(Nt), ge = Ut(kt);
class Lt extends Array {
  /**
   *
   * @param {any} initialState
   * @param {(nextState: any, state:any[], mount: boolean )=>void} mapState
   */
  constructor(t, n) {
    let s = !0;
    const o = (r) => {
      try {
        n(r, this, s);
      } finally {
        s = !1;
      }
    };
    super(void 0, o, n), o(t);
  }
  /**
   * The following code allows a mutable approach to useState
   * and useProp this with the idea of allowing an alternative
   * approach similar to Vue or Qwik of state management
   * @todo pending review with the community
   */
  // get value() {
  //     return this[0];
  // }
  // set value(nextState) {
  //     this[2](nextState, this);
  // }
}
const rt = (e) => {
  const t = Ot();
  return k(
    (n = new Lt(e, (s, o, r) => {
      s = T(s) ? s(o[0]) : s, s !== o[0] && (o[0] = s, r || t());
    })) => n
  );
}, P = (e, t) => {
  const [n] = k(([s, o, r = 0] = []) => ((!o || o && !At(o, t)) && (s = e()), [s, t, r]));
  return n;
}, _t = (e) => {
  const { current: t } = U();
  if (!(e in t))
    throw new It(
      t,
      `For useProp("${e}"), the prop does not exist on the host.`,
      e
    );
  return k(
    (n = new Lt(t[e], (s, o) => {
      s = T(s) ? s(t[e]) : s, t[e] = s;
    })) => (n[0] = t[e], n)
  );
}, C = (e, t = {}) => {
  const n = U();
  return n[e] || (n[e] = (s = t.detail) => $t(n.current, {
    type: e,
    ...t,
    detail: s
  })), n[e];
}, tt = O("atomico/options");
globalThis[tt] = globalThis[tt] || {
  sheet: !!document.adoptedStyleSheets
};
const jt = globalThis[tt], Ee = {
  checked: 1,
  value: 1,
  selected: 1
}, De = {
  list: 1,
  type: 1,
  size: 1,
  form: 1,
  width: 1,
  height: 1,
  src: 1,
  href: 1,
  slot: 1
}, Se = {
  shadowDom: 1,
  staticNode: 1,
  cloneNode: 1,
  children: 1,
  key: 1
}, _ = {}, et = [];
class nt extends Text {
}
const we = O("atomico/id"), M = O("atomico/type"), J = O("atomico/ref"), Yt = O("atomico/vnode"), ve = () => {
};
function Te(e, t, n) {
  return Bt(this, e, t, n);
}
const xt = (e, t, ...n) => {
  const s = t || _;
  let { children: o } = s;
  if (o = o ?? (n.length ? n : et), e === ve)
    return o;
  const r = e ? e instanceof Node ? 1 : (
    //@ts-ignore
    e.prototype instanceof HTMLElement && 2
  ) : 0;
  if (r === !1 && e instanceof Function)
    return e(
      o != et ? { children: o, ...s } : s
    );
  const c = jt.render || Te;
  return {
    [M]: Yt,
    type: e,
    props: s,
    children: o,
    key: s.key,
    // key for lists by keys
    // define if the node declares its shadowDom
    shadow: s.shadowDom,
    // allows renderings to run only once
    static: s.staticNode,
    // defines whether the type is a childNode `1` or a constructor `2`
    raw: r,
    // defines whether to use the second parameter for document.createElement
    is: s.is,
    // clone the node if it comes from a reference
    clone: s.cloneNode,
    render: c
  };
};
function Bt(e, t, n = we, s, o) {
  let r;
  if (t && t[n] && t[n].vnode == e || e[M] != Yt)
    return t;
  (e || !t) && (o = o || e.type == "svg", r = e.type != "host" && (e.raw == 1 ? (t && e.clone ? t[J] : t) != e.type : e.raw == 2 ? !(t instanceof e.type) : t ? t[J] || t.localName != e.type : !t), r && e.type != null && (e.raw == 1 && e.clone ? (s = !0, t = e.type.cloneNode(!0), t[J] = e.type) : t = e.raw == 1 ? e.type : e.raw == 2 ? new e.type() : o ? document.createElementNS(
    "http://www.w3.org/2000/svg",
    e.type
  ) : document.createElement(
    e.type,
    e.is ? { is: e.is } : void 0
  )));
  const c = t[n] ? t[n] : _, { vnode: f = _, cycle: a = 0 } = c;
  let { fragment: i, handlers: u } = c;
  const { children: d = et, props: l = _ } = f;
  if (u = r ? {} : u || {}, e.static && !r)
    return t;
  if (e.shadow && !t.shadowRoot && // @ts-ignore
  t.attachShadow({ mode: "open", ...e.shadow }), e.props != l && Ne(t, l, e.props, u, o), e.children !== d) {
    const m = e.shadow ? t.shadowRoot : t;
    i = Pe(
      e.children,
      /**
       * @todo for hydration use attribute and send childNodes
       */
      i,
      m,
      n,
      // add support to foreignObject, children will escape from svg
      !a && s,
      o && e.type == "foreignObject" ? !1 : o
    );
  }
  return t[n] = { vnode: e, handlers: u, fragment: i, cycle: a + 1 }, t;
}
function Ce(e, t) {
  const n = new nt(""), s = new nt("");
  let o;
  if (e[t ? "prepend" : "append"](n), t) {
    let { lastElementChild: r } = e;
    for (; r; ) {
      const { previousElementSibling: c } = r;
      if (V(r, !0) && !V(c, !0)) {
        o = r;
        break;
      }
      r = c;
    }
  }
  return o ? o.before(s) : e.append(s), {
    markStart: n,
    markEnd: s
  };
}
function Pe(e, t, n, s, o, r) {
  e = e == null ? null : re(e) ? e : [e];
  const c = t || Ce(n, o), { markStart: f, markEnd: a, keyes: i } = c;
  let u;
  const d = i && /* @__PURE__ */ new Set();
  let l = f;
  if (e && Rt(e, (m) => {
    if (typeof m == "object" && !m[M])
      return;
    const p = m[M] && m.key, y = i && p != null && i.get(p);
    l != a && l === y ? d.delete(l) : l = l == a ? a : l.nextSibling;
    const E = i ? y : l;
    let h = E;
    if (m[M])
      h = Bt(m, E, s, o, r);
    else {
      const v = m + "";
      !(h instanceof Text) || h instanceof nt ? h = new Text(v) : h.data != v && (h.data = v);
    }
    h != l && (i && d.delete(h), !E || i ? (n.insertBefore(h, l), i && l != a && d.add(l)) : E == a ? n.insertBefore(h, a) : (n.replaceChild(h, E), l = h)), p != null && (u = u || /* @__PURE__ */ new Map(), u.set(p, h));
  }), l = l == a ? a : l.nextSibling, t && l != a)
    for (; l != a; ) {
      const m = l;
      l = l.nextSibling, m.remove();
    }
  return d && d.forEach((m) => m.remove()), c.keyes = u, c;
}
function Ne(e, t, n, s, o) {
  for (const r in t)
    !(r in n) && pt(e, r, t[r], null, o, s);
  for (const r in n)
    pt(e, r, t[r], n[r], o, s);
}
function pt(e, t, n, s, o, r) {
  if (t = t == "class" && !o ? "className" : t, n = n ?? null, s = s ?? null, t in e && Ee[t] && (n = e[t]), !(s === n || Se[t] || t[0] == "_"))
    if (t[0] == "o" && t[1] == "n" && (T(s) || T(n)))
      ke(e, t.slice(2), s, r);
    else if (t == "ref")
      s && (T(s) ? s(e) : s.current = e);
    else if (t == "style") {
      const { style: c } = e;
      n = n || "", s = s || "";
      const f = R(n), a = R(s);
      if (f)
        for (const i in n)
          if (a)
            !(i in s) && bt(c, i, null);
          else
            break;
      if (a)
        for (const i in s) {
          const u = s[i];
          f && n[i] === u || bt(c, i, u);
        }
      else
        c.cssText = s;
    } else {
      const c = t[0] == "$" ? t.slice(1) : t;
      c === t && (!o && !De[t] && t in e || T(s) || T(n)) ? e[t] = s ?? "" : s == null ? e.removeAttribute(c) : e.setAttribute(
        c,
        R(s) ? JSON.stringify(s) : s
      );
    }
}
function ke(e, t, n, s) {
  if (s.handleEvent || (s.handleEvent = (o) => s[o.type].call(e, o)), n) {
    if (!s[t]) {
      const o = n.capture || n.once || n.passive ? Object.assign({}, n) : null;
      e.addEventListener(t, s, o);
    }
    s[t] = n;
  } else
    s[t] && (e.removeEventListener(t, s), delete s[t]);
}
function bt(e, t, n) {
  let s = "setProperty";
  n == null && (s = "removeProperty", n = null), ~t.indexOf("-") ? e[s](t, n) : e[t] = n;
}
const gt = {};
function x(e, ...t) {
  const n = (e.raw || e).reduce(
    (s, o, r) => s + o + (t[r] || ""),
    ""
  );
  return gt[n] = gt[n] || Oe(n);
}
function Oe(e) {
  if (jt.sheet) {
    const t = new CSSStyleSheet();
    return t.replaceSync(e), t;
  } else {
    const t = document.createElement("style");
    return t.textContent = e, t;
  }
}
const Ae = xt("host", { style: "display: contents" }), Z = O("atomico/context"), Re = (e, t) => {
  const n = U();
  ge(
    () => Mt(
      n.current,
      "ConnectContext",
      /**
       * @param {CustomEvent<import("context").DetailConnectContext>} event
       */
      (s) => {
        e === s.detail.id && (s.stopPropagation(), s.detail.connect(t));
      }
    ),
    [e]
  );
}, Me = (e) => {
  const t = C("ConnectContext", {
    bubbles: !0,
    composed: !0
  }), n = () => {
    let r;
    return t({
      id: e,
      connect(c) {
        r = c;
      }
    }), r;
  }, [s, o] = rt(
    n
  );
  return ot(() => {
    s || (e[Z] || (e[Z] = customElements.whenDefined(
      new e().localName
    )), e[Z].then(
      () => o(n)
    ));
  }, [e]), s;
}, Fe = (e) => {
  const t = Me(e), n = Ot();
  return ot(() => {
    if (t)
      return Mt(t, "UpdatedValue", n);
  }, [t]), (t || e).value;
}, Ie = (e) => {
  const t = Y(
    () => (Re(t, U().current), Ae),
    {
      props: {
        value: {
          type: Object,
          event: { type: "UpdatedValue" },
          value: () => e
        }
      }
    }
  );
  return t.value = e, t;
}, b = (e, t, n) => (t == null ? t = { key: n } : t.key = n, xt(e, t)), F = b, zt = x`*,*:before,*:after{box-sizing:border-box}button{padding:0;touch-action:manipulation;cursor:pointer;user-select:none}`, qt = x`.vh{position:absolute;transform:scale(0)}`;
function ct() {
  return D.from(/* @__PURE__ */ new Date());
}
function it(e, t = 0) {
  const n = w(e), s = n.getUTCDay(), o = (s < t ? 7 : 0) + s - t;
  return n.setUTCDate(n.getUTCDate() - o), D.from(n);
}
function Ht(e, t = 0) {
  return it(e, t).add({ days: 6 });
}
function Kt(e) {
  return D.from(new Date(Date.UTC(e.year, e.month, 0)));
}
function B(e, t, n) {
  return t && D.compare(e, t) < 0 ? t : n && D.compare(e, n) > 0 ? n : e;
}
const $e = { days: 1 };
function Ue(e, t = 0) {
  let n = it(e.toPlainDate(), t);
  const s = Ht(Kt(e), t), o = [];
  for (; D.compare(n, s) < 0; ) {
    const r = [];
    for (let c = 0; c < 7; c++)
      r.push(n), n = n.add($e);
    o.push(r);
  }
  return o;
}
function w(e) {
  return new Date(Date.UTC(e.year, e.month - 1, e.day ?? 1));
}
const Le = /^(\d{4})-(0[1-9]|1[0-2])-(0[1-9]|[12][0-9]|3[0-1])$/, W = (e, t) => e.toString().padStart(t, "0");
class D {
  constructor(t, n, s) {
    this.year = t, this.month = n, this.day = s;
  }
  // this is an incomplete implementation that only handles arithmetic on a single unit at a time.
  // i didn't want to get into more complex arithmetic since it get tricky fast
  // this is enough to serve my needs and will still be a drop-in replacement when actual Temporal API lands
  add(t) {
    const n = w(this);
    if ("days" in t)
      return n.setUTCDate(this.day + t.days), D.from(n);
    let { year: s, month: o } = this;
    "months" in t ? (o = this.month + t.months, n.setUTCMonth(o - 1)) : (s = this.year + t.years, n.setUTCFullYear(s));
    const r = D.from(w({ year: s, month: o, day: 1 }));
    return B(D.from(n), r, Kt(r));
  }
  toString() {
    return `${W(this.year, 4)}-${W(this.month, 2)}-${W(this.day, 2)}`;
  }
  toPlainYearMonth() {
    return new z(this.year, this.month);
  }
  equals(t) {
    return D.compare(this, t) === 0;
  }
  static compare(t, n) {
    return t.year < n.year ? -1 : t.year > n.year ? 1 : t.month < n.month ? -1 : t.month > n.month ? 1 : t.day < n.day ? -1 : t.day > n.day ? 1 : 0;
  }
  static from(t) {
    if (typeof t == "string") {
      const n = t.match(Le);
      if (!n)
        throw new TypeError(t);
      const [, s, o, r] = n;
      return new D(
        parseInt(s, 10),
        parseInt(o, 10),
        parseInt(r, 10)
      );
    }
    return new D(
      t.getUTCFullYear(),
      t.getUTCMonth() + 1,
      t.getUTCDate()
    );
  }
}
class z {
  constructor(t, n) {
    this.year = t, this.month = n;
  }
  add(t) {
    const n = w(this), s = (t.months ?? 0) + (t.years ?? 0) * 12;
    return n.setUTCMonth(n.getUTCMonth() + s), new z(n.getUTCFullYear(), n.getUTCMonth() + 1);
  }
  equals(t) {
    return this.year === t.year && this.month === t.month;
  }
  toPlainDate() {
    return new D(this.year, this.month, 1);
  }
}
function st(e, t) {
  if (t)
    try {
      return e.from(t);
    } catch {
    }
}
function I(e) {
  const [t, n] = _t(e);
  return [P(() => st(D, t), [t]), (r) => n(r.toString())];
}
function _e(e) {
  if (e) {
    const [t, n] = e.split("/"), s = st(D, t), o = st(D, n);
    if (s && o)
      return [s, o];
  }
  return [];
}
function je(e) {
  const [t, n] = _t(e);
  return [P(() => _e(t), [t]), (r) => n(`${r[0]}/${r[1]}`)];
}
function $(e, t) {
  return P(
    () => new Intl.DateTimeFormat(t, { timeZone: "UTC", ...e }),
    [t, e]
  );
}
function Et(e, t, n) {
  const s = $(e, n);
  return P(() => {
    const o = [], r = /* @__PURE__ */ new Date();
    for (var c = 0; c < 7; c++) {
      const f = (r.getDay() - t + 7) % 7;
      o[f] = s.format(r), r.setDate(r.getDate() + 1);
    }
    return o;
  }, [t, s]);
}
const Dt = (e, t, n) => B(e, t, n) === e, St = (e) => e.target.matches(":dir(ltr)"), Ye = { month: "long", day: "numeric" }, xe = { month: "long" }, Be = { weekday: "narrow" }, ze = { weekday: "long" }, X = { bubbles: !0 };
function qe({ props: e, context: t }) {
  const { offset: n } = e, {
    firstDayOfWeek: s,
    isDateDisallowed: o,
    min: r,
    max: c,
    page: f,
    locale: a,
    focusedDate: i
  } = t, u = ct(), d = Et(ze, s, a), l = Et(Be, s, a), m = $(Ye, a), p = $(xe, a), y = P(
    () => f.start.add({ months: n }),
    [f, n]
  ), E = P(
    () => Ue(y, s),
    [y, s]
  ), h = C("focusday", X), v = C("selectday", X), q = C("hoverday", X);
  function lt(g) {
    h(B(g, r, c));
  }
  function Gt(g) {
    let S;
    switch (g.key) {
      case "ArrowRight":
        S = i.add({ days: St(g) ? 1 : -1 });
        break;
      case "ArrowLeft":
        S = i.add({ days: St(g) ? -1 : 1 });
        break;
      case "ArrowDown":
        S = i.add({ days: 7 });
        break;
      case "ArrowUp":
        S = i.add({ days: -7 });
        break;
      case "PageUp":
        S = i.add(g.shiftKey ? { years: -1 } : { months: -1 });
        break;
      case "PageDown":
        S = i.add(g.shiftKey ? { years: 1 } : { months: 1 });
        break;
      case "Home":
        S = it(i, s);
        break;
      case "End":
        S = Ht(i, s);
        break;
      default:
        return;
    }
    lt(S), g.preventDefault();
  }
  function Qt(g) {
    const S = y.equals(g);
    if (!S && !t.showOutsideDays)
      return;
    const Vt = g.equals(i), ut = g.equals(u), ft = w(g), L = o?.(ft), dt = !Dt(g, r, c);
    let ht = "", A;
    if ("highlightedRange" in t) {
      const [H, K] = t.highlightedRange, mt = H?.equals(g), yt = K?.equals(g);
      A = H && K && Dt(g, H, K), ht = `${mt ? "range-start" : ""} ${yt ? "range-end" : ""} ${A && !mt && !yt ? "range-inner" : ""}`;
    } else
      A = t.value?.equals(g);
    return {
      part: `${`button day ${// we don't want outside days to ever be shown as selected
      S ? A ? "selected" : "" : "outside"} ${L ? "disallowed" : ""} ${ut ? "today" : ""}`} ${ht}`,
      tabindex: S && Vt ? 0 : -1,
      disabled: dt,
      "aria-disabled": L ? "true" : void 0,
      "aria-pressed": S && A,
      "aria-current": ut ? "date" : void 0,
      "aria-label": m.format(ft),
      onkeydown: Gt,
      onclick() {
        L || v(g), lt(g);
      },
      onmouseover() {
        !L && !dt && q(g);
      }
    };
  }
  return {
    weeks: E,
    yearMonth: y,
    daysLong: d,
    daysShort: l,
    formatter: p,
    getDayProps: Qt
  };
}
const G = ct(), at = Ie({
  firstDayOfWeek: 1,
  isDateDisallowed: () => !1,
  focusedDate: G,
  page: { start: G.toPlainYearMonth(), end: G.toPlainYearMonth() }
});
customElements.define("calendar-month-ctx", at);
const He = Y(
  (e) => {
    const t = Fe(at), n = se(), s = qe({ props: e, context: t });
    function o() {
      n.current.querySelector("button[tabindex='0']")?.focus();
    }
    return /* @__PURE__ */ F("host", { shadowDom: !0, focus: o, children: [
      /* @__PURE__ */ b("div", { id: "h", part: "heading", children: s.formatter.format(w(s.yearMonth)) }),
      /* @__PURE__ */ F("table", { ref: n, "aria-labelledby": "h", part: "table", children: [
        /* @__PURE__ */ b("thead", { children: /* @__PURE__ */ b("tr", { part: "tr head", children: s.daysLong.map((r, c) => /* @__PURE__ */ F("th", { part: "th", scope: "col", children: [
          /* @__PURE__ */ b("span", { class: "vh", children: r }),
          /* @__PURE__ */ b("span", { "aria-hidden": "true", children: s.daysShort[c] })
        ] })) }) }),
        /* @__PURE__ */ b("tbody", { children: s.weeks.map((r, c) => /* @__PURE__ */ b("tr", { part: "tr week", children: r.map((f, a) => {
          const i = s.getDayProps(f);
          return /* @__PURE__ */ b("td", { part: "td", children: i && /* @__PURE__ */ b("button", { ...i, children: f.day }) }, a);
        }) }, c)) })
      ] })
    ] });
  },
  {
    props: {
      offset: {
        type: Number,
        value: 0
      }
    },
    styles: [
      zt,
      qt,
      x`:host{--color-accent: black;--color-text-on-accent: white;display:flex;flex-direction:column;gap:.25rem;text-align:center;inline-size:fit-content}table{border-collapse:collapse;font-size:.875rem}th{font-weight:700;block-size:2.25rem}td{padding-inline:0}button{color:inherit;font-size:inherit;background:transparent;border:0;font-variant-numeric:tabular-nums;block-size:2.25rem;inline-size:2.25rem}button:hover:where(:not(:disabled)){background:#0000000d}button:is([aria-pressed=true],:focus-visible){background:var(--color-accent);color:var(--color-text-on-accent)}button:focus-visible{outline:1px solid var(--color-text-on-accent);outline-offset:-2px}button:disabled,:host::part(outside),:host::part(disallowed){cursor:default;opacity:.5}`
    ]
  }
);
customElements.define("calendar-month", He);
function wt(e) {
  return /* @__PURE__ */ b(
    "button",
    {
      part: `button ${e.name} ${e.onclick ? "" : "disabled"}`,
      onclick: e.onclick,
      "aria-disabled": e.onclick ? null : "true",
      children: /* @__PURE__ */ b("slot", { name: e.name, children: e.children })
    }
  );
}
function Jt(e) {
  const t = w(e.page.start), n = w(e.page.end);
  return /* @__PURE__ */ F("div", { role: "group", "aria-labelledby": "h", part: "container", children: [
    /* @__PURE__ */ b("div", { id: "h", class: "vh", "aria-live": "polite", "aria-atomic": "true", children: e.formatVerbose.formatRange(t, n) }),
    /* @__PURE__ */ F("div", { part: "header", children: [
      /* @__PURE__ */ b(wt, { name: "previous", onclick: e.previous, children: "Previous" }),
      /* @__PURE__ */ b("div", { part: "heading", "aria-hidden": "true", children: e.format.formatRange(t, n) }),
      /* @__PURE__ */ b(wt, { name: "next", onclick: e.next, children: "Next" })
    ] }),
    /* @__PURE__ */ b(
      at,
      {
        value: e,
        onselectday: e.onSelect,
        onfocusday: e.onFocus,
        onhoverday: e.onHover,
        children: /* @__PURE__ */ b("slot", {})
      }
    )
  ] });
}
const Zt = {
  value: {
    type: String,
    value: ""
  },
  min: {
    type: String,
    value: ""
  },
  max: {
    type: String,
    value: ""
  },
  isDateDisallowed: {
    type: Function,
    value: (e) => !1
  },
  firstDayOfWeek: {
    type: Number,
    value: () => 1
  },
  showOutsideDays: {
    type: Boolean,
    value: !1
  },
  locale: {
    type: String,
    value: () => {
    }
  },
  months: {
    type: Number,
    value: 1
  },
  focusedDate: {
    type: String,
    value: () => {
    }
  }
}, Wt = [
  zt,
  qt,
  x`:host{display:block;inline-size:fit-content}[role=group]{display:flex;flex-direction:column;gap:1em}:host::part(header){display:flex;align-items:center;justify-content:space-between}:host::part(heading){font-weight:700;font-size:1.25em}button{display:flex;align-items:center;justify-content:center}button[aria-disabled]{cursor:default;opacity:.5}`
], Ke = { year: "numeric" }, Je = { year: "numeric", month: "long" };
function vt(e, t) {
  return (t.year - e.year) * 12 + t.month - e.month;
}
const Tt = (e, t) => (e = t === 12 ? new z(e.year, 1) : e, {
  start: e,
  end: e.add({ months: t - 1 })
});
function Xt({
  months: e,
  locale: t,
  focusedDate: n,
  setFocusedDate: s
}) {
  const [o] = I("min"), [r] = I("max"), c = C("focusday"), f = C("change"), a = P(
    () => B(n ?? ct(), o, r),
    [n, o, r]
  ), [i, u] = rt(
    () => Tt(a.toPlainYearMonth(), e)
  ), d = (h) => {
    const v = vt(i.start, h.toPlainYearMonth());
    return v >= 0 && v < e;
  };
  ot(() => {
    let h = i.start;
    if (!d(a)) {
      const v = vt(h, a.toPlainYearMonth()), q = Math.floor(v / e);
      h = h.add({ months: q * e });
    }
    u(Tt(h, e));
  }, [a, e]);
  const l = U();
  function m() {
    l.current.querySelectorAll("calendar-month").forEach((h) => h.focus());
  }
  function p(h) {
    s(h), c(w(h));
  }
  const y = $(Ke, t), E = $(Je, t);
  return {
    format: y,
    formatVerbose: E,
    page: i,
    focusedDate: a,
    dispatch: f,
    onFocus(h) {
      h.stopPropagation(), p(h.detail), setTimeout(m);
    },
    min: o,
    max: r,
    next: r == null || !d(r) ? () => p(a.add({ months: e })) : void 0,
    previous: o == null || !d(o) ? () => p(a.add({ months: -e })) : void 0,
    focus: m
  };
}
const Ze = Y(
  (e) => {
    const [t, n] = I("value"), [s = t, o] = I("focusedDate"), r = Xt({
      ...e,
      focusedDate: s,
      setFocusedDate: o
    });
    function c(f) {
      n(f.detail), r.dispatch();
    }
    return /* @__PURE__ */ b("host", { shadowDom: !0, focus: r.focus, children: /* @__PURE__ */ b(
      Jt,
      {
        ...e,
        ...r,
        value: t,
        onSelect: c
      }
    ) });
  },
  { props: Zt, styles: Wt }
);
customElements.define("calendar-date", Ze);
const Ct = (e, t) => D.compare(e, t) < 0 ? [e, t] : [t, e], We = Y(
  (e) => {
    const [t, n] = je("value"), [s = t[0], o] = I("focusedDate"), r = Xt({
      ...e,
      focusedDate: s,
      setFocusedDate: o
    }), c = C("rangestart"), f = C("rangeend"), [a, i] = rt();
    function u(p) {
      r.onFocus(p), d(p);
    }
    function d(p) {
      p.stopPropagation(), i((y) => y && { ...y, b: p.detail });
    }
    function l(p) {
      const y = p.detail;
      p.stopPropagation(), a ? (n(Ct(a.a, y)), i(void 0), f(w(y)), r.dispatch()) : (i({ a: y, b: y }), c(w(y)));
    }
    const m = a ? Ct(a.a, a.b) : t;
    return /* @__PURE__ */ b("host", { shadowDom: !0, focus: r.focus, children: /* @__PURE__ */ b(
      Jt,
      {
        ...e,
        ...r,
        highlightedRange: m,
        onFocus: u,
        onHover: d,
        onSelect: l
      }
    ) });
  },
  { props: Zt, styles: Wt }
);
customElements.define("calendar-range", We);
export {
  Ze as CalendarDate,
  He as CalendarMonth,
  We as CalendarRange
};
